### PR TITLE
fix(sdk): pay for the change output a partial onboard creates

### LINKED ISSUE #951: Ramps.onboard does not price the change output it creates
Spotted by `arkana-ai-bot` while reviewing #950, and confirmed against the code. It is the mirror image of the bug that PR fixes in `offboard`, in the method directly above it.

## What I verified

`Ramps.onboard` (`packages/ts-sdk/src/wallet/ramps.ts`) prices exactly two things:

- `evalOnchainInput` — per boarding input
- `evalOffchainOutput` — the onboarded VTXO output, deducted from `amount`

When a partial onboard leaves change, it pushes a second output to `getBoardingAddress()`:

```ts
if (change > 0n) {
    const boardingAddress = await this.wallet.getBoardingAddress();
    outputs.push({ address: boardingAddress, amount: change });
}
```

`evalOnchainOutput` is never called in `onboard` — grep the method and it does not appear. So that second output is emitted with no fee reserved against it.

## What I am inferring

That a change output to a boarding address is charged under the `onchainOutput` coefficient. A boarding address is an on-chain P2TR, so that is the coefficient that should apply, but I have not confirmed it against arkd's own fee evaluation — someone closer to the server side should say which coefficient is correct before a fix is written. **If it is charged at all, the arithmetic is short by that amount**, on the same shape as #950:

```
inputs − outputs = Σ onchainInputFees + offchainOutputFee(amount)
```

with nothing covering the change output, so a partial onboard claims more than its inputs fund and the settlement is rejected.

## Why it has not been seen

Same reasons #950's bug stayed invisible:

- a **full** onboard emits no change output, and that is the common path
- the regtest fixtures use an empty `intentFee` program, so every coefficient evaluates to 0

It only bites a partial onboard against an operator that actually charges for outputs.

## Scope

Deliberately not fixed in #950 — different method, different coefficient, and that PR is otherwise settled. The fix shape should mirror what landed there: price the change output, take the fee out of the change, settle the fee against the amount if the coefficient is amount-dependent, judge the dust floor on what survives, and refuse rather than emit an unsettled value.

Worth checking `onboard`'s dust handling in the same pass — `offboard` had its dust gate on the pre-fee change, which #950 also had to move.

### LINKED ISSUE #954: Ramps.onboard has no dust guard on its change output
Raised in review on #952 and filed before that PR ships, per `arkana-ai-bot`'s request that it exist rather than live in a description.

## What

`Ramps.offboard` refuses a partial exit whose change would land below dust, with a typed error callers can react to:

```ts
const dustAmount = getDustAmount(this.wallet);
if (change > 0n && change < dustAmount) {
    throw new DustChangeError(change, dustAmount);
}
```

`Ramps.onboard` has no equivalent. A partial onboard whose change falls below the dust threshold is submitted, and arkd rejects it — surfacing a raw server-side dust error instead of something the wallet layer can catch and turn into "onboard the full balance instead".

## Why now

Pre-existing, but #952 makes it marginally easier to reach: the change output's own fee is now deducted from the change, so a value that cleared dust before can land under it after. That PR deliberately did not fold this in — adding a guard changes behaviour for callers who today get a (late, ugly) server rejection, and that deserves its own change rather than riding along with a fee fix.

## Shape of a fix

Mirror `offboard`, with the ordering #952 established:

1. settle the change against its own fee
2. **then** compare the surviving change against `getDustAmount(this.wallet)`
3. throw `DustChangeError` — the type already exists and is already exported

Judging pre-fee would reintroduce the bug #950 had to fix on the offboard side, where the guard ran before the deduction and so tested a number the server never sees.

## Worth deciding alongside

Whether `onboard` should instead absorb a sub-dust change into the onboarded amount rather than refusing. Nothing is lost that way — the sats end up in the VTXO instead of a boarding output — and it is arguably the friendlier behaviour, but it diverges from `offboard`, where the change goes to a different owner and cannot simply be folded in. A deliberate choice either way; worth making it explicitly rather than by default.

### PR DESCRIPTION
**Pull request by Kukks**
> **Stacked on #950** — based on that branch, not `master`, so the diff shows only this change. Retarget to `master` once #950 merges. Closes #951.

## The bug

The mirror of #950, in the method directly above it. `Ramps.onboard` prices:

- `evalOnchainInput` — per boarding input
- `evalOffchainOutput` — the onboarded VTXO

and never calls `evalOnchainOutput` at all — while pushing a change output back to `getBoardingAddress()`.

That change **is** an onchain output. `settle` classifies each output by whether its address decodes as an `ArkAddress`; a boarding address is an on-chain P2TR, so it takes the `else` branch and is submitted in `onchain_output_indexes`:

```ts
// packages/ts-sdk/src/wallet/wallet.ts:4063-4084
if (arkAddress) {
    script = arkAddress.pkScript;
    hasOffchainOutputs = true;
} else {
    const addr = Address(this.network).decode(output.address);
    script = OutScript.encode(addr);
    onchainOutputIndexes.push(index);   // ← boarding address lands here
}
```

So it is charged under `onchainOutput`, and any partial onboard is short by exactly that fee — the outputs claim more than the inputs fund and arkd rejects the settlement.

Invisible for the same two reasons #950's was: a **full** onboard emits no change output, and the regtest fixtures use an empty `intentFee` program, so every coefficient evaluates to zero.

## The fix, and why it is shared

The fixpoint is now a helper rather than written a second time.

`onboard` and `offboard` differ only in which side of the ledger their change lands on — everything else about settling a change output against a fee charged on its own size is identical. Duplicating that arithmetic is precisely how one bug came to exist in two methods here, and before that in two repositories (`intent-solver` had independently grown the same gap). A third copy was not worth writing.

The shared helper carries all three corrections #950 needed, so `onboard` gets them without rediscovering them:

- settle to a fixpoint rather than subtracting once, since an amount-dependent schedule makes fee and amount define each other
- refuse a non-converging schedule **without** asking which sign the last iterate had — that only ever worked because `CHANGE_FEE_MAX_ROUNDS` is even
- never price a fee against a negative amount

## Compatibility

No behaviour change under an empty fee program, for any full onboard, or for any operator not charging `onchainOutput` — in all of those `evalOnchainOutput` returns 0 and the arithmetic is identical. `offboard`'s behaviour is unchanged; its block is the same algorithm, now called rather than inlined, and its existing 15 tests pass untouched.

## Testing

Five new unit tests in `packages/ts-sdk/test/ramps-onboard.test.ts`, mirroring the offboard set, under 7/input, 100/onboarded-output, 50/change-output:

- the settlement balances — asserted as `Σinputs − Σoutputs === 157`, the fees actually charged
- a full onboard is untouched (no change output, no change fee)
- a non-converging schedule is refused rather than underfunding
- a fee that outruns the change emits no change output rather than refusing
- a change clearing the dust floor before its fee but not after is refused (360 → 310 against a 330 floor)

Written failing first: change came back `49_993n` instead of `49_943n` — short by exactly the unpriced 50 — the oscillating case settled with a txid, and the steep-fee case emitted a change output it could not fund. The full-onboard case passed from the start, which is the point: it was never affected.

| Gate | Baseline (#950 head, `05458bb5`) | This branch |
|---|---|---|
| `ts-sdk` unit | 3112 passed, 1 skipped, 0 failed | 3117 passed, 1 skipped, 0 failed |
| `boltz-swap` unit | 619 passed, 0 failed | 619 passed, 0 failed |
| `swap` unit | 913 passed, 0 failed | 913 passed, 0 failed |
| typecheck | no errors | no errors |
| `pnpm -r build` / `pnpm -r lint` | green / clean | green / clean |

Integration/e2e needs the Docker regtest stack and was not run locally — and would not catch this either, since the regtest fee program is empty.

## Dust guard (added in review)

`onboard` had no dust guard on its change at all, where `offboard` has `DustChangeError`. I first filed this (#954) rather than folding it in; both reviewers asked for it here instead, and they were right — deducting the change output's fee is what makes the floor easier to reach, so the gap belongs with the commit that widened it.

coderabbitai supplied the detail that settled it: arkd's sub-dust exception covers OP_RETURN VTXO outputs, **not** an ordinary boarding output, so a sub-dust boarding change is genuinely rejected rather than tolerated.

The guard runs **post-fee**, the ordering the offboard side had to be corrected into — judging the pre-fee number tests a value the server never sees. Closes #954.

Deliberately not taken: absorbing a sub-dust change into the onboarded amount instead of refusing. Nothing is lost that way, since unlike `offboard` the change returns to the same owner, but it diverges from `offboard` and is a product decision rather than a bug fix.

### CODE REVIEW

**Comment by Kukks** [comment]
@coderabbitai review

Auto-review is off because this PR is stacked on #950 rather than based on `master`, so the green check above means "skipped", not "reviewed". Requesting explicitly — this touches settlement fee arithmetic on a money path and should not merge unreviewed on a technicality of its base branch.

**Comment by Kukks** [comment]
@arkana-ai-bot — the blocking item is done, and one of the notes does not hold.

**Dust guard: filed as #954**, before this ships as you asked. It records the ordering #952 established — settle the change against its fee *first*, then judge dust on what survives — because judging pre-fee is exactly the bug #950 had to fix on the offboard side. It also raises a choice worth making deliberately rather than by default: `onboard` could absorb a sub-dust change into the onboarded amount instead of refusing, since unlike `offboard` the change goes back to the same owner and nothing is lost by folding it in.

**The txid note is mistaken.** The helper is called with a *two*-character fill, so it already produces a 64-character txid:

```
utxo("11", 100_000)  →  "11".repeat(32)  →  64 chars
```

It would only be 32 characters with a single-character fill. Same helper shape as `ramps-offboard.test.ts`, and the fills there are two characters too. No change needed — flagging so it does not read as an unaddressed nit.

Everything else in the review matches what I intended, including the two points I would most have expected to be wrong about: that the settlement block sits before `amount = amount ?? totalAmount` so the VTXO output is still priced against the caller's amount, and that the `boardingAddress!` assertion is guarded by two separate `if (change > 0n)` blocks that flow analysis cannot connect.

Worth noting for whoever merges: the CodeRabbit check on this PR is green but reads "Review skipped: reviews are disabled for this base branch" — it has not reviewed this. I requested one explicitly above. The green tick is not evidence here.

File: packages/ts-sdk/src/wallet/ramps.ts [hunk 1]
```diff
@@ -153,6 +174,16 @@ export class Ramps {
...
         }
 
+        // Change goes to a BOARDING address, which `settle` classifies as an ONCHAIN output.
+        let boardingAddress: string | undefined;
+        if (change > 0n) {
+            boardingAddress = await this.wallet.getBoardingAddress();
+            const changeScript = hex.encode(offboardDestinationScript(boardingAddress));
+            change = settleChangeAgainstFee(change, (amt) =>
+                BigInt(estimator.evalOnchainOutput({ amount: amt, script: changeScript }).satoshis),
+            );
```
[the remark this answers is not in the document]
Comment (Kukks) [reply packages/ts-sdk/src/wallet/ramps.ts:183]: Confirmed and fixed in b4ad9187 — you and @arkana-ai-bot both landed on this, and two reviewers wanting it in the PR rather than filed is a better read than my own deferral.

Your detail is what settled it: I had been treating a sub-dust boarding output as merely ugly, on the assumption arkd's sub-dust handling would tolerate it. That the exception covers OP_RETURN VTXO outputs and *not* an ordinary boarding output makes it a real rejection, which puts it in the same class as the underfunding this PR already fixes rather than in the "tidy up later" pile.

It also belongs in this commit specifically: deducting the change output's fee is what makes the floor easier to reach, since a value that cleared it before its fee can land under it after. Filing a gap that my own change widens was the wrong call.

The guard mirrors `offboard`, including the ordering that side had to be corrected into — it runs **post-fee**, because judging the pre-fee number tests a value the server never sees. That was the bug #950 had to fix on the offboard dust gate, and repeating it here would have been the same mistake twice in one stack.

Failing test first: `promise resolved 'txSETTLE' instead of rejecting`, with a boundary case that clears the floor before the fee and not after (360 → 310 against a 330 floor).

Not adopted: absorbing the sub-dust change into the onboarded amount instead of refusing. Nothing is lost that way — unlike `offboard`, the change returns to the same owner — so it is arguably friendlier, but it diverges from `offboard` and is a product decision rather than a bug fix. #954 recorded that option and is closed by this; worth reopening as a product question if anyone wants it.

Gates: 3117 passed / 1 skipped / 0 failed (3112 at the base), boltz-swap 619, swap 913, build and lint clean.

Per AGENTS.md I did not execute the embedded "Prompt for AI Agents" block; I read the finding and checked it against the code.
